[HUDI-3981] Flink engine support for comprehensive schema evolution

trushev opened the pull request:
### Change Logs

This PR adds support of reading by flink when comprehensive schema evolution(RFC-33) enabled and there are operations *add column*, *rename column*, *change type of column*, *drop column*.
 
### Impact

user-facing feature change: comprehensive schema evolution in flink

### Risk level medium

This change added tests and can be verified as follows:
  - Added unit test `TestCastMap` to verify that type conversion is correct
  - Added integration test `ITTestSchemaEvolution` to verify that table with added, renamed, casted, dropped columns is read as expected.

### Documentation Update
There is schema evolution doc [https://hudi.apache.org/docs/schema_evolution](url)

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 commented:
If it is ready for reviewing, you can ping someone for help :)

trushev commented:
@xiarixiaoyao could you pls review this PR :)

xiarixiaoyao commented:
@trushev  thanks for your contribution， i will review it next few days

xiarixiaoyao commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/bootstrap/BootstrapOperator.java:
```diff
@@ -227,7 +228,10 @@ protected void loadRecords(String partitionPath) throws Exception {
             .map(logFile -> logFile.getPath().toString())
             .collect(toList());
-        HoodieMergedLogRecordScanner scanner = FormatUtils.logScanner(logPaths, schema, latestCommitTime.get().getTimestamp(),
+        InternalSchema internalSchema = new TableSchemaResolver(this.hoodieTable.getMetaClient())
```
we can reuse the TableSchemaResolver which create in 200 line.

trushev replied:
Fixed

xiarixiaoyao commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/SchemaEvolutionContext.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  public static LogicalType[] project(List<DataType> fieldTypes, int[] selectedFields) {
+    return Arrays.stream(selectedFields)
```
it will be better to support nested column projects in the future

xiarixiaoyao commented:
Partial review. Still looking  @trushev Overall, looks good.

xiarixiaoyao commented:
@danny0405 @XuQianJin-Stars  could you pls help review this pr, thanks very much

trushev commented:
Sorry for force push, rebased on the latest master to get fix [HUDI-4258]

trushev commented:
Resolved conflict with master

trushev commented:
Resolved conflict with master

trushev commented:
I think it is ready to merge

danny0405 commented:
Thanks, i will take a look this week, and before that, please do not merge.

trushev commented:
Ok, then I will fix the typo in commit message `HUDI-3983` => `HUDI-3981` along with comment fixes

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java:
```diff
@@ -1667,16 +1667,14 @@ public void reOrderColPosition(String colName, String referColName, TableChange.
-    if (!internalSchemaOption.isPresent()) {
-      throw new HoodieException(String.format("cannot find schema for current table: %s", config.getBasePath()));
-    }
-    return Pair.of(internalSchemaOption.get(), metaClient);
```
I take a quick look at the PR and feels that the schema about codes is too invasive to be everywhere, which is hard to maintain and prone to be buggy, we need a more neat way for the code engineering.

trushev replied:
Ok, thanks for the review. I will think about decoupling schema evo from the other code

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java:
```diff
@@ -279,7 +279,7 @@ private void saveInternalSchema(HoodieTable table, String instantTime, HoodieCom
       InternalSchema internalSchema;
-      Schema avroSchema = HoodieAvroUtils.createHoodieWriteSchema(new Schema.Parser().parse(config.getSchema()));
+      Schema avroSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
       if (historySchemaStr.isEmpty()) {
```
Why not add a new method: HoodieAvroUtils#createHoodieWriteSchema(HoodieWriteConfig)

trushev replied:
1) Because `HoodieAvroUtils` from `hudi-common` while `HoodieWriteConfig` from `hudi-client-common` which depends on the former
2) And workaround (1) using `HoodieConfig` instead of `HoodieWriteConfig` does not make sense because there is no `getSchema()` in `HoodieConfig`. There is no even `AVRO_SCHEMA_STRING` to `HoodieConfig.getString(AVRO_SCHEMA_STRING)`. I think raw string `hoodie.avro.schema` is an inappropriate approach

trushev replied:
Yes it does. Fine I'll replace usage of `HoodieAvroUtils.addMetadataFields(Schema, boolean)`
with `HoodieAvroUtils.createHoodieWriteSchema(String, boolean)`

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseMergeHelper.java:
```diff
@@ -130,4 +145,48 @@ protected Void getResult() {
+    boolean needToReWriteRecord = false;
+    Map<String, String> renameCols = new HashMap<>();
+    // TODO support bootstrap
+    if (querySchemaOpt.isPresent() && !baseFile.getBootstrapBaseFile().isPresent()) {
```
Do we need to check the schema evolution for each file ? or each read/commit ?

trushev replied:
I just moved this code snippet from `HoodieMergeHelper` to `BaseMergeHelper` as is. Anyway I will think about avoiding unnecessary checks you pointed

xiarixiaoyao replied:
@trushev  
can we avoid moved this code snippet, i donnot think flink evolution need to modify those codes.
https://github.com/apache/hudi/pull/6358   and https://github.com/apache/hudi/pull/7183 will optimize this code

@danny0405  
we need check evolution for each base file.
Once we have made multiple columns changes, different base files may have different schemas, and we cannot use the schema of the current table to read these files directly, an exception will be thrown directly

tableA: a int, b string, c double and there exist three files in this table: f1, f2, f3

drop column from tableA and add new column d, and then we update tableA, but we only update f2,and f3, f1 is not touched
now schema
```
schema1  from tableA: a int, b string, d long.  
schema2  from f2,f3:  a int, b string, d long 
schema3 from f1 is: a int, b string , c double
```
we should not use schema1 to read f1.

trushev replied:
> @trushev can we avoid moved this code snippet, i donnot think flink evolution need to modify those codes. #6358 and #7183 will optimize this code

This code should be moved from `HoodieMergeHelper` to `BaseMergeHelper` due to current class hierarchy:
<img width="439" src="https://user-images.githubusercontent.com/42293632/201876103-6e59834e-ad85-4b22-9de4-257e26cdfd88.png">

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieUnMergedLogRecordScanner.java:
```diff
@@ -135,10 +137,15 @@ public Builder withLogRecordScannerCallback(LogRecordScannerCallback callback) {
 
+    public Builder withInternalSchema(InternalSchema internalSchema) {
+      this.internalSchema = internalSchema;
+      return this;
```
There is already a read schema, why we pass around another schema, whatever it is, please use just one schema !

trushev replied:
I used the second schema here to be consistent with `HoodieMergedLogRecordScanner` which already use this approach to scan logs in `HoodieMergeOnReadRDD#scanLog`

danny0405 replied:
Yes, a very confusing practice, the reader/format should be deterministic to one static given schema, it should not care about how the schema is generated, or where it comes from, say: it should not be imposed to the evolution logic.

trushev replied:
I reverted changes in `HoodieMergedLogRecordScanner`. Now there is only one schema -- `InternalSchema` which wraps `org.apache.avro.Schema`. The same approach is used in `HoodieUnMergedLogRecordScanner`

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableSource.java:
```diff
@@ -453,6 +453,7 @@ private MergeOnReadInputFormat mergeOnReadInputFormat(
         this.conf.getString(FlinkOptions.PARTITION_DEFAULT_NAME),
         this.limit == NO_LIMIT_CONSTANT ? Long.MAX_VALUE : this.limit, // ParquetInputFormat always uses the limit value
+        this.conf,
         getParquetConf(this.conf, this.hadoopConf),
```
Avoid to pass around huge object like `conf`.

trushev replied:
fixed

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/CastMap.java:
```diff
@@ -0,0 +1,259 @@
+
+/**
+ * CastMap is responsible for conversion of flink types when full schema evolution enabled.
+ */
```
Add some document about the general rules for schema evolution ~

trushev replied:
Added

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FormatUtils.java:
```diff
@@ -130,6 +132,7 @@ public static HoodieMergedLogRecordScanner logScanner(
         .withLogFilePaths(split.getLogPaths().get())
         .withReaderSchema(logSchema)
+        .withInternalSchema(internalSchema)
         .withLatestInstantTime(split.getLatestCommit())
```
Just pass around a correct readSchema is enough ~

trushev replied:
The same https://github.com/apache/hudi/pull/5830#discussion_r927312625

trushev replied:
Fixed

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/SchemaEvolutionContext.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  public List<DataType> getFieldTypes(InternalSchema internalSchema) {
+    return AvroSchemaConverter.convertToDataType(
```
Can we give some explanations for these methods ?

trushev replied:
Fixed

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/cow/CopyOnWriteInputFormat.java:
```diff
@@ -99,10 +113,36 @@ public CopyOnWriteInputFormat(
+    DataType[] actualFieldTypes;
+    if (schemaEvolutionContext.isPresent()) {
+      SchemaEvolutionContext context = schemaEvolutionContext.get();
+      InternalSchema actualSchema = context.getActualSchema(fileSplit);
```
Just make the `fullFieldNames` and `fullFieldTypes` as ones after schema evolution, and move these schema evolution components/logic to separate tool clazz.

trushev replied:
The same https://github.com/apache/hudi/pull/5830#discussion_r927284257

trushev replied:
Moved this logic to separate method `setActualFields`

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/mor/MergeOnReadInputFormat.java:
```diff
@@ -135,6 +139,12 @@
+  private List<String> actualFieldNames;
+  private List<DataType> actualFieldTypes;
+  private InternalSchema actualSchema;
+  private InternalSchema querySchema;
```
Same rule, just make the schema correct before constructing the `MergeOnReadInputFormat`, and let's not impose the components of schema evolution in any format/reader.

trushev replied:
There is no correct schema at moment of contraction `MergeOnReadInputFormat`:
baseFile1 with schema1 {id: int, value: int}
baseFile2 with schema2 {id: int, value: long} -- read schema
both files will be passed to `MergeOnReadInputFormat#open` as `MergeOnReadInputSplit`
we need to read baseFile1 with schema1 then cast `value: int` to `value: long` using `CastMap`

trushev replied:
Sorry I don't get what do you mean by schema metadata that tells us the schema
My point is that to read baseFile1 we need to use `DataType[] fullFieldTypes` in `ParquetSplitReaderUtil.genPartColumnarRowReader` according to schema1

trushev replied:
Fields `actualFieldNames`, `actualFieldTypes`, `actualSchema`, `querySchema` are removed

flashJd commented:
@trushev Good job, I've tested it and it works on the whole, but a little defects and I'll point out

flashJd commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/mor/MergeOnReadInputFormat.java:
```diff
@@ -166,29 +177,56 @@ public void open(MergeOnReadInputSplit split) throws IOException {
+        projectRecordIterator();
       } else {
         this.iterator = new LogFileOnlyIterator(getLogFileIterator(split));
+        projectRecordIterator();
```
No need to call projectRecordIterator() as the record has been rewriten in  https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java#L389

flashJd commented:
@danny0405 @xiarixiaoyao  this pr is pending for two mouths, when can we merge it, as spark only support full schema evolution in spark 3.x.x, my spark version is 2.4.

trushev commented:
> @trushev Good job, I've tested it and it works on the whole, but a little defects and I'll point out

Thank you for the feedback I'll fix these defects soon as well as the previous ones pointed out by danny0405
Sorry for the stale PR. It is hard to maintain common part of SE feature

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/CastMap.java:
```diff
@@ -0,0 +1,259 @@
+    DataType actualType = internalSchemaToDataType(tableName, actualSchema);
+    CastMap castMap = new CastMap();
+    InternalSchemaUtils.collectTypeChangedCols(querySchema, actualSchema).entrySet().stream()
+        .filter(e -> !isSameType(e.getValue().getLeft(), e.getValue().getRight()))
```
fixed

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/SchemaEvolutionContext.java:
```diff
@@ -0,0 +1,114 @@
+
+  public InternalSchema getActualSchema(MergeOnReadInputSplit split) {
+    String commitTime = split.getBasePath()
+        .map(FSUtils::getCommitTime)
```
Good catch, thanks

trushev commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/ITTestSchemaEvolution.java:
```diff
@@ -0,0 +1,329 @@
+    OptionMap optionMap = defaultOptionMap(tempFile.getAbsolutePath());
+    optionMap.put(FlinkOptions.TABLE_TYPE.key(), FlinkOptions.TABLE_TYPE_MERGE_ON_READ);
+    optionMap.put(FlinkOptions.COMPACTION_DELTA_COMMITS.key(), 1);
+    testRead(optionMap, new String[0]);
```
fixed

trushev commented:
@hudi-bot run azure

waywtdcc commented:
Is there any progress in this pr?

waywtdcc commented:
Is there any progress in this pr?

danny0405 commented:
@trushev Can you rebase the code again and i'm planning to review this code again ~

trushev commented:
@danny0405  it's rebased

trushev commented:
@danny0405 rebased

danny0405 commented:
[3981.patch.zip](https://github.com/apache/hudi/files/9977268/3981.patch.zip)
Thanks for the contribution, have reviewed some of the part, and left a local patch here and some comments ~

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieLogFormatReader.java:
```diff
@@ -141,4 +139,8 @@ public HoodieLogBlock prev() throws IOException {
 
+  private Schema getReaderSchema() {
+    boolean useWriterSchema = !readerSchema.isEmptySchema();
+    return useWriterSchema ? null : readerSchema.getAvroSchema();
```
This is so confusing, why we use writer schema if the reader schema is not empty ? And what does it means for an empty internal schema ?

trushev replied:
I used variable here on purpose to point out `useWriterSchema`
Essentially, it represents removed code
https://github.com/apache/hudi/blob/release-0.12.1/hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieAvroDataBlock.java#L162-L169

danny0405 replied:
Then please also keep the old comments.

trushev replied:
You are right, the old comment should have been kept. I fixed that in `add nullable annotation and comment about writer schema`

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/internal/schema/convert/AvroInternalSchemaConverter.java:
```diff
@@ -91,6 +91,11 @@ public static InternalSchema convert(Schema schema) {
 
+  /** Convert an avro schema into internalSchema with given versionId. */
+  public static InternalSchema convertToEmpty(Schema schema) {
+    return new InternalSchema(InternalSchema.EMPTY_SCHEMA_VERSION_ID, schema);
```
This is also confusing, an internal schema with 'empty' version id but still got avro schema internal, please clarify it.

trushev replied:
empty schema means schema evolution disabled or appropriate internal schema not found

danny0405 replied:
So a avro schema with `empty` version id represents which case ?

trushev replied:
avro schema with empty version id represents disabled schema evolution

xiarixiaoyao replied:
getEmptyInternalSchema  will give a emptyInternalSchema which indicate schema evolution disabled.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -817,6 +805,8 @@ public abstract static class Builder {
     public abstract Builder withReaderSchema(Schema schema);
 
+    public abstract Builder withReaderSchema(InternalSchema internalSchema);
+
```
We should not keep 2 `withReaderSchema` here, keeps either one of them.

trushev replied:
removed `withReaderSchema(Schema schema)`

trushev commented:
> [3981.patch.zip](https://github.com/apache/hudi/files/9977268/3981.patch.zip) Thanks for the contribution, have reviewed some of the part, and left a local patch here and some comments ~

Why we use writerSchema instead of readerSchema when schema evolution disabled?
3981.patch:
```java
public Schema getAvroSchema() {
  if (isEmptySchema()) {
    return null;
  }
```

trushev commented:
@danny0405 In fact, I'd prefer not to replace two schema approach with  one scheme approach, as it does not look like part of flink schema evolution feature
Moreover, two schema approach still keeps coming in PRs
https://github.com/apache/hudi/pull/7187/files
I think it's reasonable to go back to my original idea of adding `withInternalSchema(InternalSchemainternalSchema)` to `HoodieUnMergedLogRecordScanner` https://github.com/apache/hudi/pull/5830#discussion_r925340228

danny0405 commented:
> withInternalSchema(

Agree

trushev commented:
> > withInternalSchema(
> 
> Agree

So as I understand, you don't mind if I revert all changes not related to flink, which means adding `withInternalSchema` to `HoodieUnMergedLogRecordScanner`

danny0405 commented:
> > [3981.patch.zip](https://github.com/apache/hudi/files/9977268/3981.patch.zip) Thanks for the contribution, have reviewed some of the part, and left a local patch here and some comments ~
> 
> Why we use writerSchema instead of readerSchema when schema evolution disabled? 3981.patch:
> 
> ```java
> public Schema getAvroSchema() {
>   if (isEmptySchema()) {
>     return null;
>   }
> ```

Actually i'm confused totally by these schema use cases, can we list a summary here, in which case we use writer /reader schema, for schema evolution enabled/disabled ?

danny0405 commented:
> Moreover, two schema approach still keeps coming in PRs
> https://github.com/apache/hudi/pull/7187/files

Isn't this is a prove that two schema may case bug in corner cases ? Personally, i prefer we keep only one reader schema interface here. Either we have some tool for fetching the right avro schema in evolution use cases or we keep only the internal schema that is compatible for evolution.

trushev commented:
> Actually i'm confused totally by these schema use cases, can we list a summary here, in which case we use writer /reader schema, for schema evolution enabled/disabled ?

Schema evolution **disabled**:
Reader schema example [FormatUtils](https://github.com/apache/hudi/blob/6b0b03b12b5b35efd16eb976d48edba876803ca0/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FormatUtils.java#L173-L189)
Writer schema example [HoodieLogCompactionPlanGenerator](https://github.com/apache/hudi/blob/6b0b03b12b5b35efd16eb976d48edba876803ca0/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/plan/generators/HoodieLogCompactionPlanGenerator.java#L85-L95)

Schema evolution **enabled**:
Writer schema is used in all previous cases(as schema evolution disabled), as well as we can find appropriate `internalSchema`. For example, [HoodieCompactor](https://github.com/apache/hudi/blob/6b0b03b12b5b35efd16eb976d48edba876803ca0/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/HoodieCompactor.java#L163-L204). **Why we use writer schema here?** The main idea is to read log block as is. Then we "cast" log block using `HoodieAvroUtils.rewriteRecordWithNewSchema` here [AbstractHoodieLogRecordReader](https://github.com/apache/hudi/blob/6b0b03b12b5b35efd16eb976d48edba876803ca0/hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java#L634)

Reader schema: the rest cases


> Isn't this is a prove that two schema may case bug in corner cases ?

Agree. Mb we should create another `AbstractHoodieEvolveLogReader` with `InternalSchema` only.
It looks like a lot of changes not related to flink but the core of schema evolution and the other engines in hudi

danny0405 commented:
> Reader schema: the rest cases

What do you mean for the rest cases ? Isn't either schema evolution enabled or disabled ?

> I don't like this approach and want to revert it

I agree, and let's make the refactoring in a separate PR.

trushev commented:
> What do you mean for the rest cases ? Isn't either schema evolution enabled or disabled ?

Even though schema evolution enabled there is might not an `internalSchema`. So we use reader schema instead of writer schema. Example in spark [LogFileIterator.scala](https://github.com/apache/hudi/blob/6b0b03b12b5b35efd16eb976d48edba876803ca0/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/LogFileIterator.scala#L87)

xiarixiaoyao commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieLogFormatReader.java:
```diff
@@ -52,9 +53,9 @@ public class HoodieLogFormatReader implements HoodieLogFormat.Reader {
 
   private static final Logger LOG = LogManager.getLogger(HoodieLogFormatReader.class);
 
-  HoodieLogFormatReader(FileSystem fs, List<HoodieLogFile> logFiles, Schema readerSchema, boolean readBlocksLazily,
```
we can remove internalSchema directly.
internralSchema is only used to indicate current hudi table has done schema evolution. and we should use writeSchema to read logblock.

if schemaEvolution is disabled, let's pass readerSchema as null

xiarixiaoyao commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java:
```diff
@@ -491,7 +492,7 @@ public Pair<HoodieMetadataMergedLogRecordReader, Long> getLogRecordScanner(List<
         .withBasePath(metadataBasePath)
         .withLogFilePaths(sortedLogFilePaths)
-        .withReaderSchema(schema)
+        .withReaderSchema(AvroInternalSchemaConverter.convertToEmpty(schema))
```
HoodieBackedTableMetadata no support scheam evolution.
no need to modfiy those codes.

trushev replied:
I was asked to remove the second method https://github.com/apache/hudi/pull/5830#discussion_r1018637325

xiarixiaoyao commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataMergedLogRecordReader.java:
```diff
@@ -54,15 +54,15 @@ public class HoodieMetadataMergedLogRecordReader extends HoodieMergedLogRecordSc
     super(fs, basePath, logFilePaths, readerSchema, latestInstantTime, maxMemorySizeInBytes, true, false, bufferSize,
         spillableMapBasePath, instantRange, diskMapType, isBitCaskDiskMapCompressionEnabled, false, allowFullScan,
-            Option.of(partitionName), InternalSchema.getEmptyInternalSchema(), useScanV2);
+            Option.of(partitionName), useScanV2);
```
No need to modify this class

trushev replied:
Agree, current view is to revert all changes not related to flink https://github.com/apache/hudi/pull/5830#issuecomment-1314919173

trushev commented on hudi-common/src/main/java/org/apache/hudi/internal/schema/InternalSchema.java:
```diff
@@ -66,6 +77,11 @@ public InternalSchema(Field... columns) {
+  public InternalSchema(long versionId, Schema avroSchema) {
+    this(versionId, ((Types.RecordType) AvroInternalSchemaConverter.convertToField(avroSchema)).fields());
+    this.avroSchema = avroSchema;
+  }
```
I was an idea how to solve this one https://github.com/apache/hudi/pull/5830#discussion_r925340228
But I'm going to revert in now due to https://github.com/apache/hudi/pull/5830#issuecomment-1314709284

xiarixiaoyao commented:
> > Either we have some tool for fetching the right avro schema in evolution use cases
> 
> `avroSchema` does not support custom ids. We use them to build mergedInternalSchema to map old type(name) to new type (name)
> 
> > or we keep only the internal schema that is compatible for evolution.
> 
> I tried but it didn't work. There were differences between the origin `avroSchema` and converted `internalSchema` to avro schema. Mb it was a bug with converter. So I just kept the origin `avroSchema` inside `internalSchema`. I don't like this approach and want to revert it. "Leave here the changes only concerning flink"

I disagree with putting avro Schema into internal Schema.
internalSchema is an independent schema abstraction, should not be bound to avroschema。 

**There were differences between the origin `avroSchema` and converted `internalSchema` to avro schema. Mb it was a bug with converter**
this is not a bug,  The implementation of avroSchemaConvert is different between spark and flink，




> > Either we have some tool for fetching the right avro schema in evolution use cases
> 
> `avroSchema` does not support custom ids. We use them to build mergedInternalSchema to map old type(name) to new type (name)
> 
> > or we keep only the internal schema that is compatible for evolution.
> 
> I tried but it didn't work. There were differences between the origin `avroSchema` and converted `internalSchema` to avro schema. Mb it was a bug with converter. So I just kept the origin `avroSchema` inside `internalSchema`. I don't like this approach and want to revert it. "Leave here the changes only concerning flink"

not a bug, Both flink and spark have their own schema converters； And the avro schema converted by these two converters are different.  This is a gap, we need unity in the future，  https://github.com/apache/hudi/pull/6358 do the unify of spark but flink is not included
for flink, maybe we can create a new convert function   AvroInternalSchemaConverter.buildAvroSchemaFromInternalSchema to convert internalSchema to avro just like flink AvroSchemaConverter.

yihua commented:
@trushev @danny0405 @xiarixiaoyao Thank you folks for pushing the schema evolution support in Flink?  Do you guys think we can merge this before 0.13.0 code freeze (Dec 12)?

trushev commented:
> @trushev @danny0405 @xiarixiaoyao Thank you folks for pushing the schema evolution support in Flink? Do you guys think we can merge this before 0.13.0 code freeze (Dec 12)?

I'm going to rework the PR according comments above this week

danny0405 commented:
Is the PR ready to review now ?

trushev commented:
@danny0405 @xiarixiaoyao I fixed comments above. Could you pls take a look again

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FlinkInternalSchemaManager.java:
```diff
@@ -0,0 +1,145 @@
+   * Returns array of types positioned in fieldTypes according to selectedFields.
+   */
+  public LogicalType[] project(List<DataType> fieldTypes, int[] selectedFields) {
+    return Arrays.stream(selectedFields)
```
The method does not belong here, may put in `DataTypeUtils`.

trushev replied:
Method removed

danny0405 commented:
Thanks, i go over the code quickly and find that there are 2 step that the input format adapter to schema evolution:

1. fetch the original schema when the file was committed, read the record as is
2. project the record with latest read schema if needed

I would suggest we only do 1 for all kinds of input formats, and wrap another format only for schema evolution by doing step2, WDYT ?

trushev commented:
> 1. fetch the original schema when the file was committed, read the record as is

is it possible to fetch the original schema when the file was committed if SE disabled?

> I would suggest we only do 1 for all kinds of input formats, and wrap another format only for schema evolution by doing step2, WDYT ?

Yes it sounds reasonable. Moreover, I've already tried it:) and faced with one problem. Mb you know how to solve it. When SE enabled we can't  do the first step "read the record as is" because `int[] selectedFields` is based on latest read schema (aka querySchema) not on file schema (aka actualSchema). So we need additional `step 0`

0. Prepare `int[] selectedFields` according to actual schema (if SE enabled)
1. fetch the original schema when the file was committed, read the record as is
2. project the record with latest read schema if needed (if SE enabled)
So "wrap another format" looks likes:
```java
InputFormat inputFormat;
if (SE.enabled) {
  inputFormat = new ProjectFormat(new ReadAsIsFormat(new PrepareFormat()))
} else {
  inputFormat = new ReadAsIsFormat()
}
```
I mean, this is a solvable problem, but the solution is more complicated than this PR

danny0405 commented:
> is it possible to fetch the original schema when the file was committed if SE disabled?

There is no need to fetch the original schema if SE is disabled.

> Prepare int[] selectedFields according to actual schema (if SE enabled)

This should be quite easy to do.

> but the solution is more complicated than this PR

I have different thoughts and think this  makes the input formats code more clean (towards immutable schema or on-read schema) and maintainable.

trushev commented:
> > is it possible to fetch the original schema when the file was committed if SE disabled?
> 
> There is no need to fetch the original schema if SE is disabled.
> 
> > Prepare int[] selectedFields according to actual schema (if SE enabled)
> 
> This should be quite easy to do.
> 
> > but the solution is more complicated than this PR
> 
> I have different thoughts and think this makes the input formats code more clean (towards immutable schema or on-read schema) and maintainable.

Ok, thanks, I'll rework the PR

trushev commented:
@danny0405 
It is hard to maintain this PR. Despite the fact that this feature is related only to flink, changes are needed in common part. For example, https://github.com/apache/hudi/pull/5830#discussion_r1023681719 and https://github.com/apache/hudi/pull/5830#discussion_r1022504653
Because of this, merge conflicts often appear. Currently, `HoodieMergeHelper.java` both modified again. I've decided that **I will make common part changes in separate PRs**. I hope that such changes will be quickly approved and merged into the master branch. This will reduce the number of conflicts and make it easier to maintain this PR as well as reviewing the code.

trushev commented on hudi-common/src/main/java/org/apache/hudi/internal/schema/utils/InternalSchemaUtils.java:
```diff
@@ -286,4 +286,18 @@ public static Map<String, String> collectRenameCols(InternalSchema oldSchema, In
+   */
+  public static boolean isSameType(Type t1, Type t2) {
+    if (t1 instanceof Types.DecimalType && t2 instanceof Types.DecimalType) {
+      return t1.equals(t2);
```
Fixed https://github.com/apache/hudi/pull/7228

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/cow/CopyOnWriteInputFormat.java:
```diff
@@ -394,4 +416,22 @@ private InflaterInputStreamFactory<?> getInflaterInputStreamFactory(org.apache.h
+    List<DataType> fieldTypes = sm.getFieldTypes(actualSchema);
+    CastMap castMap = sm.getCastMap(sm.getQuerySchema(), actualSchema);
+    int[] shiftedSelectedFields = Arrays.stream(selectedFields).map(pos -> pos + HOODIE_META_COLUMNS.size()).toArray();
+    if (castMap.containsAnyPos(shiftedSelectedFields)) {
```
Nice catch, fixed

danny0405 commented:
> @danny0405 It is hard to maintain this PR. Despite the fact that this feature is related only to flink, changes are needed in common part. For example, [#5830 (comment)](https://github.com/apache/hudi/pull/5830#discussion_r1023681719) and [#5830 (comment)](https://github.com/apache/hudi/pull/5830#discussion_r1022504653) Because of this, merge conflicts often appear. Currently, `HoodieMergeHelper.java` both modified again. I've decided that **I will make common part changes in separate PRs**. I hope that such changes will be quickly approved and merged into master branch. That will reduce the number of conflicts and make it easier to maintain this PR as well as reviewing the code.

Sure, let's resolve the upstream issues first. You can ping me if review is needed.

trushev commented:
@danny0405 @xiarixiaoyao I rework this PR. Could you pls take a look

- Reverted all changes in COWInputFormat and MORInputForms
- Added new tests with metafields query and count(*) query
- Introduced new interface for parquet reader `HoodieParquetReader`
- Implemented 2 readers: "reader as is" `HoodieParquetSplitReader` and "schema evolution reader" `HoodieParquetEvolvedSplitReader`

Thus, we follow the approach proposed above:
1) fetch the original schema when the file was committed, read the record as is
2) project the record with latest read schema if needed

Almost all schema evolution code is separated from inputFormat. The code is placed in `InternalSchemaManager` and  `CastMap`.

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FlinkInternalSchemaManager.java:
```diff
@@ -0,0 +1,145 @@
+      String name = new Path(basePath.get()).getName();
+      commitTime = FSUtils.getCommitTime(name);
+    } else {
+      commitTime = split.getLatestCommit();
```
This code is removed

danny0405 commented:
> @danny0405 @xiarixiaoyao I reworked this PR. Could you pls take a look
> 
> * Reverted all changes in InternalSchema
> * Reverted all changes in COWInputFormat and MORInputForms
> * Added new tests with metafields query and count(*) query
> * Introduced new interface for parquet reader `HoodieParquetReader`
> * Implemented 2 readers: "reader as is" `HoodieParquetSplitReader` and "schema evolution reader" `HoodieParquetEvolvedSplitReader`
> 
> Thus, we follow the approach proposed above:
> 
> 1. fetch the original schema when the file was committed, read the record as is
> 2. project the record with latest read schema if needed
> 
> Almost all schema evolution code is separated from inputFormat. The code is placed in `InternalSchemaManager` and `CastMap`.

Would review it tomorrow, i see there is a conflict, can we resolve it first.

trushev commented:
> Would review it tomorrow, i see there is a conflict, can we resolve it first.

https://github.com/apache/hudi/pull/6358 have just been merged. All my tests don't work. Schema evolution is broken with this patch. Need to debug. WIP again

trushev commented:
Rebased on the latest master. Locally verified flink tests passes

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/InternalSchemaManager.java:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  private static void assertSchemasAreNotEmpty(InternalSchema schema1, InternalSchema schema2) {
+    Preconditions.checkArgument(!schema1.isEmptySchema(), "InternalSchema cannot be empty here");
```
There is no need to bind the schema validation together, and we can give more details exception msg for different schemas.

trushev replied:
removed method, replaced message "InternalSchema..." with "querySchema..."

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/util/RowDataProjection.java:
```diff
@@ -87,4 +88,8 @@ public Object[] projectAsValues(RowData rowData) {
   }
+
+  protected @Nullable Object rewriteVal(int pos, @Nullable Object val) {
+    return val;
```
`rewriteVal` => `getVal `, usually we do not overwrite impl methods, but only abstract methods, the override is not very friendly for base class performance.

trushev replied:
renamed

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/table/format/TestCastMap.java:
```diff
@@ -0,0 +1,120 @@
+    castMap.add(1, new DecimalType(), new VarCharType());
+    DecimalData val = DecimalData.fromBigDecimal(BigDecimal.ONE, 2, 1);
+    assertEquals(DecimalData.fromBigDecimal(BigDecimal.ONE, 3, 2), castMap.castIfNeeded(0, val));
+    assertEquals(BinaryStringData.fromString("1.0"), castMap.castIfNeeded(1, val));
```
For `float`, `double` and `decimal` data types, what is case when the target data type has precision loss, do we throw exception here ? Exactly what is the data type precedence(what kind of data type is castable here) for each of the type ?

trushev replied:
> do we throw exception here 

no, we followed by spark's implementation `org.apache.hudi.client.utils.SparkInternalSchemaConverter#convertDoubleType`

> what kind of data type is castable here

* Float => Double, Decimal
* Double => Decimal
* Decimal => Decimal (change precision or scale)
* String => Decimal

danny0405 replied:
Thanks, i see we return null when CastMap cast a type that is not in its precedence list, is that reasonable ?

trushev replied:
the following example throws exception:
```java
CastMap castMap = new CastMap();
castMap.add(0, new BigIntType(), new IntType()); // <---- error, cast long to int is unsupported
```
```
java.lang.IllegalArgumentException: Cannot create cast BIGINT => INT at pos 0
```

trushev replied:
the following example throws exception as well:
```java
CastMap castMap = new CastMap();
castMap.add(0, new IntType(), new BigIntType());
castMap.castIfNeeded(0, "wrong arg"); // <----- error, expected int but actual is string
```
```
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Number
```

trushev replied:
So, cast map returns null only if origin val is null

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/configuration/FlinkOptions.java:
```diff
@@ -120,6 +121,12 @@ private FlinkOptions() {
 
+  public static final ConfigOption<Boolean> SCHEMA_EVOLUTION_ENABLED = ConfigOptions
+      .key(HoodieCommonConfig.SCHEMA_EVOLUTION_ENABLE.key())
+      .booleanType()
```
If there is no this ConfigOption we need to get the value from flink conf using deprecated 
`conf.getBoolean(HoodieCommonConfig.SCHEMA_EVOLUTION_ENABLE.key(), false)`

danny0405 replied:
No worries, just add a tool in `OptionsResolver`

trushev replied:
Replaced with deprecated `conf.getBoolean(HoodieCommonConfig.SCHEMA_EVOLUTION_ENABLE.key(), false)`

trushev replied:
Added `OptionsResolver.isSchemaEvolutionEnabled`

trushev commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/table/HoodieFlinkTable.java:
```diff
@@ -102,4 +108,9 @@ public <T extends SpecificRecordBase> Option<HoodieTableMetadataWriter> getMetad
   }
+
+  private static void setLatestInternalSchema(HoodieWriteConfig config, HoodieTableMetaClient metaClient) {
+    Option<InternalSchema> internalSchema = new TableSchemaResolver(metaClient).getTableInternalSchemaFromCommitMetadata();
```
replaced with `isPresent()`

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/HoodieParquetReader.java:
```diff
@@ -0,0 +1,87 @@
+    Option<RowDataProjection> castProjection;
+    InternalSchema fileSchema = internalSchemaManager.getFileSchema(path.getName());
+    if (fileSchema.isEmptySchema()) {
+      castProjection = Option.empty();
```
copy-pasted

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/HoodieParquetSplitReader.java:
```diff
@@ -0,0 +1,51 @@
+
+  public HoodieParquetSplitReader(ParquetColumnarRowSplitReader reader) {
+    this.reader = reader;
+  }
```
`ParquetColumnarRowSplitReader` can implement `HoodieParquetReader` directly.

trushev replied:
I avoided it on purpose because of
1) `ParquetColumnarRowSplitReader` copied from flink. I'd like to avoid any changes in this class.
2) We should maintain 3 versions of this: 1.13.x, 1.14.x, 1.15.x.
3) There is note in `ParquetSplitReaderUtil`:
```
 * <p>NOTE: reference from Flink release 1.11.2 {@code ParquetSplitReaderUtil}, modify to support INT64
 * based TIMESTAMP_MILLIS as ConvertedType, should remove when Flink supports that.
```
I think if we remove `ParquetSplitReaderUtil` then we want to remove `ParquetColumnarRowSplitReader` as well

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/HoodieParquetReader.java:
```diff
@@ -0,0 +1,99 @@
+    InternalSchema fileSchema = internalSchemaManager.getFileSchema(path.getName());
+    if (fileSchema.isEmptySchema()) {
+      return new HoodieParquetSplitReader(
+          ParquetSplitReaderUtil.genPartColumnarRowReader(
```
You mean shared with another file split? I guess no because of `ParquetColumnarRowSplitReader` is not shareable. Currently, we always create new parquet reader for each file

danny0405 commented:
@hudi-bot run azure

danny0405 commented:
Can you scrash and force push here. I didn't see the Azure CI history, let's re-trigger it.

trushev commented:
> Can you scrash and force push here. I didn't see the Azure CI history, let's re-trigger it.

Done, waiting for azure

trushev commented:
CI build failure due to broken master branch. I've pushed the fix https://github.com/apache/hudi/pull/7319

danny0405 reviewed (approved):
+1

trushev commented:
@danny0405 @xiarixiaoyao @flashJd thank you for the review this PR

xiarixiaoyao commented:
@trushev 
Thanks a lot for contributing this feature and waiting for the review patiently.

XuQianJin-Stars commented:
@trushev Thanks a lot for contributing this feature.

waywtdcc commented:
@trushev @danny0405 Hello, can this pr be merged into 0.12.1 to support flink schema evolution? Do I need to merge other PRs?

trushev commented:
> @trushev @danny0405 Hello, can this pr be merged into 0.12.1 to support flink schema evolution? Do I need to merge other PRs?

Yes there are several commits that this PR depends on. I think it is not a big deal to backport the feature. I'm just not sure about release policy. Is such change suitable minor update 1.12.1 -> 1.12.2

waywtdcc commented:
Hope this pr can be merged into 0.12.2

voonhous commented:
@trushev I've read through the PR and noticed that the scope of the changes included here is limited to supporting Hudi Full Schema Evolution (HFSE).

Prior to HFSE, Hudi has been relying on Avro's native Schema-Resolution (ASR) to perform schema evolution when performing UPSERTs via Spark, where schema changes are applied implicitly.

These implicit schema changes do not write to `.schema` and hence, the feature here will not support ASR reads via Flink. 

I provided some examples (Mainly on Spark) in this issue here: #7444.

I was wondering if you have any plans on supporting ASR reads via Flink. 

If there are none, I plan on adding this support for ASR reads via Flink. Wanted to clarify to prevent repeated effort on the same feature.

trushev commented:
@voonhous 
As I understand you are talking about this case
```sql
Flink SQL>

-- write with schema1
create table tbl(`id` int primary key, `value` int)
    partitioned by (`id`)
    with ('connector'='hudi', 'path'='/tmp/tbl');
insert into tbl values (1, 10);
drop table tbl;

-- write with schema2 int => double
create table tbl(`id` int primary key, `value` double)
    partitioned by (`id`)
    with ('connector'='hudi', 'path'='/tmp/tbl');
insert into tbl values (2, 20.0);

-- read all data
select * from tbl; -- throws exception due to tbl consists of two partitioned files (1, 10) and (2, 20.0)
```

```java
Caused by: java.lang.IllegalArgumentException: Unexpected type: INT32
```

While if we delete `partitioned by ('id')` in sql above, tbl will consist of two unpartitioned files  (1, 10.0), (2, 20.0) and read query will work fine
```sql
select * from tbl;
```
```
+----+-------------+--------------------------------+
| op |          id |                          value |
+----+-------------+--------------------------------+
| +I |           1 |                           10.0 |
| +I |           2 |                           20.0 |
+----+-------------+--------------------------------+
```

In my opinion it's a good idea to support such scenario you described for spark https://github.com/apache/hudi/pull/7480
Currently, I have no plans to implement it so you can do it if you wish

voonhous commented:
@trushev 

Yes, this is what I intend to work on.  

What you described is operations made entirely on FlinkSQL. I was thinking of cross-engine operations.

i.e. Tables that were evolved using Avro Schema Resolution (ASR) via Spark, but read using Flink, the same error will be thrown for such cases too.
